lib/logstash/plugin_mixins/ecs_compatibility_support/target_check.rb

```diff
@@ -63,17 +69,28 @@ module RegisterDecorator
           # @override
           def register
             super.tap do
-              if target_set? && codec_target_set?
-                # both is not good. prefer codec.
-                logger.warn(MULTIPLE_TARGETS_MESSAGE)
-              elsif target_set? && (codec_target_set? == false)
-                # prefer codec's target option
-                logger.info(PREFER_CODEC_TARGET_MESSAGE +
-                              "To do so, remove the `target` directive from this plugin and instead define the codec with " +
-                              "`codec => #{codec.config_name} { target => #{target.inspect} }`")
-              elsif (target_set? == false) && codec_target_set?.nil?
-                # target desired but unset, codec doesn't provide
-                logger.info(TARGET_NOT_SET_MESSAGE)
+              plugin_targeted = target_set?
+              codec_targeted = codec_target_set?
+
+              if plugin_targeted
+                if codec_targeted
+                  # both is not good. prefer codec.
+                  logger.warn(MULTIPLE_TARGETS_MESSAGE)
+                elsif codec_targeted == false
+                  # prefer codec's target option
+                  logger.info(PREFER_CODEC_TARGET_MESSAGE +
+                                "To do so, remove the `target` directive from this plugin and instead define the codec with " +
+                                "`codec => #{codec.config_name} { target => #{target.inspect} }`")
+                end
+              elsif (plugin_targeted == false)
+                # ECS enabled and plugin not targeted
+                if codec_targeted.nil?
+                  # codec does not support target
+                  logger.info(ECS_TARGET_NOT_SET_MESSAGE)
+                elsif codec_targeted == false
+                  # codec supports codec, but it is unspecified
+                  logger.info(ECS_TARGET_NOT_SET_PREFER_CODEC_MESSAGE)
```

Comment:
This is an interesting one. Those two suggestions you made have skipped the !target && !codec.target, while my original print the msg when both are not set. I thought this is your key idea that not to recommend user to set target because it is not consistent with other suggestions :facepalm: .Until Logstash core managed to pass config from input to codec, we are not recommended setting target if both are unset

Reply:
If both are unset, we only push the recommendation when ECS is enabled and are otherwise silent. This allows a user who explicitly opts out of ECS to avoid getting repeatedly informed about something that they don't care about:

~~~
     input
       when configured with a codec that supports `target`
         and neither the input's target nor the codec's target is set
           and ECS compatibility is enabled
             logs info advocating for setting the codec's target
           and ECS compatibility is disabled
             does not log info about targets
~~~